ci-operator/step-registry/rosa/cluster/upgrade/hcp/rosa-cluster-upgrade-hcp-commands.sh

```diff
@@ -0,0 +1,226 @@
+  start_time=$(date +"%s")
+  while true; do
+    if (( $(date +"%s") - $start_time >= 1800 )); then
+      log "error: Timed out while waiting for the previous upgrade schedule to be removed."
+      exit 1
+    fi
+
+    rosa upgrade cluster -y -m auto --version $upgrade_version -c $cluster_id  1>"/tmp/update_info.txt" 2>&1 || true
+    upgrade_info=$(cat "/tmp/update_info.txt")
+    if [[ "$upgrade_info" == *"There is already"* ]]; then
+      log "Waiting for the previous upgrade schedule to be removed."
+      sleep 120
+    else
+      log -e "$upgrade_info"
+      break
+    fi
+  done
+
+  # Monitor cluster
+  start_time=$(date +"%s")
+  while true; do
+    sleep 120
+    log "Wait for the cluster upgrading finished ..."
+    current_version=$(rosa describe cluster -c $cluster_id -o json | jq -r '.openshift_version')
+    if [[ "$current_version" == "$upgrade_version" ]]; then
+      record_cluster "timers.ocp_upgrade" "${upgrade_version}" $(( $(date +"%s") - "${start_time}" ))
+      log "Upgrade the cluster $cluster_id to the openshift version $upgrade_version successfully after $(( $(date +"%s") - ${start_time} )) seconds"
+      break
+    fi
+
+    if (( $(date +"%s") - $start_time >= $CLUTER_UPGRADE_TIMEOUT )); then
+      log "error: Timed out while waiting for the cluster upgrading to be ready"
+      set_proxy
+      oc get clusteroperators
+      exit 1
+    fi
+  done
+
+}
+
+function upgrade_machinepool_to () {
+  rosa list machinepool -c $cluster_id
+
+  up_version=$1
+  mp_id_list=$(rosa list machinepool -c $cluster_id -o json | jq -r ".[].id")
+  for mp_id in $mp_id_list; do
+    log "rosa upgrade machinepool $mp_id -y -c $cluster_id --version $up_version"
+    rosa upgrade machinepool $mp_id -y -c $cluster_id --version $up_version
+  done
+
+  for mp_id in $mp_id_list; do
+    start_time=$(date +"%s")
+    while true; do
+        sleep 120
+        log "Wait for the node upgrading for the machinepool $mp_id finished ..."
+        node_version=$(rosa list machinepool -c $cluster_id -o json | jq -r --arg k $mp_id '.[] | select(.id==$k) .version.id')
+        if [[ "$node_version" =~ ${up_version}- ]]; then
+          record_cluster "timers.machinset_upgrade" "${mp_id}" $(( $(date +"%s") - "${start_time}" ))
+          log "Upgrade the machinepool $mp_id successfully to the openshift version $up_version after $(( $(date +"%s") - ${start_time} )) seconds"
+          break
+        fi
+
+        if (( $(date +"%s") - $start_time >= $NODE_UPGRADE_TIMEOUT )); then
+          log "error: Timed out while waiting for the machinepool upgrading to be ready"
+          rosa list machinepool -c $cluster_id
+          exit 1
+        fi
+    done
+  done
+}
+
+
+read_profile_file() {
+  local file="${1}"
+  if [[ -f "${CLUSTER_PROFILE_DIR}/${file}" ]]; then
+    cat "${CLUSTER_PROFILE_DIR}/${file}"
+  fi
+}
+
+# Log in
+SSO_CLIENT_ID=$(read_profile_file "sso-client-id")
```

Comment:
What's this? SSO login? Can this be moved to prow_ci.sh?

Reply:
It is ROSA client login flag for OpenID client